File: Examples/ButtonsExample/UILabel+Theming.m
//
//  UILabel+Theming.m
//  ButtonsExample
//
//  Created by Eric Horacek on 1/1/16.
//  Copyright © 2016 Eric Horacek. All rights reserved.
//

@import Motif;

#import "ThemeSymbols.h"

#import "UILabel+Theming.h"

@implementation UILabel (Theming)

+ (void)load {
    [self
        mtf_registerThemeProperties:@[
            ThemeProperties.fontName,
            ThemeProperties.fontSize
        ]
        requiringValuesOfType:@[
            NSString.class,
            NSNumber.class
        ]
        applierBlock:^(NSDictionary<NSString *, id> *properties, UILabel *label, NSError **error) {
            NSString *name = properties[ThemeProperties.fontName];
            CGFloat size = [properties[ThemeProperties.fontSize] floatValue];
            UIFont *font = [UIFont fontWithName:name size:size];

            if (font != nil) {
                label.font = font;
                return YES;
            }

            return [self
                mtf_populateApplierError:error
                withDescriptionFormat:@"Unable to create a font named %@ of size %@", name, @(size)];
        }];
}

@end

File: Examples/SwiftButtonsExample/UILabel+Theming.swift
//
//  UILabel+Theming.swift
//  SwiftButtonsExample
//
//  Created by Eric Horacek on 4/7/15.
//  Copyright (c) 2015 Eric Horacek. All rights reserved.
//

import Foundation
import Motif

extension UILabel {
    open override class func initialize() {
        guard self === UILabel.self else { return }
        
        self.mtf_registerThemeProperties([
                ThemeProperties.fontName.rawValue,
                ThemeProperties.fontSize.rawValue
            ],
            requiringValuesOfType: [
                NSString.self,
                NSNumber.self
            ],
            applierBlock: { (properties, label, error) -> Bool in
                guard let name = properties[ThemeProperties.fontName.rawValue] as? String else { return false }
                guard let size = properties[ThemeProperties.fontSize.rawValue] as? NSNumber else { return false }
                guard let label = label as? UILabel else { return false }

                let font = UIFont(name: name, size: CGFloat(size.floatValue))

                if let font = font {
                    label.font = font
                    return true
                }

                return self.mtf_populateApplierError(error, withDescription: "Unable to create font named \(name) of size \(size)")
            })
    }
}
